### PR TITLE
Add support for updated file versions. Ensure compatibility.

### PR DESCRIPTION
**Pull request by pihart**
Perform all file operations in `files.ts`.

### CODE REVIEW

File: src/lib/qboard.ts [hunk 1]
```diff
@@ -212,10 +216,8 @@ export default class QBoard {
     iEvent.e.preventDefault();
     this.updateCursor(iEvent);
     this.dragLeave();
-    const imgs = await this.pages.processFiles(
-      (iEvent.e as DragEvent).dataTransfer.files
-    );
```
Comment (pihart) [inline src/lib/qboard.ts:217]: This command was duplicated before

**Comment by pihart** [comment]
Mostly addresses #36; do whatever you want to induce an error instead of
```js
default:
  return data;
```

I strongly recommend deprecating file format version 0 right now and then breaking compatibility in the near future. It was just recently introduced so I don't see that as an issue. You can upgrade the format just by opening and saving. (And people would be able to recover by just adding a little boilerplate around the existing file.

Once we do that, we should be able to verify qboard json files (aside from accidental or malicious corruption) just by checking if the field `qboard-version` exists.

File: src/lib/files.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+// manages version compatibility with old document formats
+// change the signature and usages to accommodate new data; this will fill in sample data for missing fields
+export class JSONReader {
+  static async read(reader: AsyncReader): Promise<PageJSON[]> {
+    const json = (await reader).result.toString();
+    const object = JSON.parse(json);
+
+    // V0
+    if (Array.isArray(object)) return object;
```
Comment (cjquines) [inline src/lib/files.ts:29]: do we really need to check this

Comment (cjquines) [reply src/lib/files.ts:29]: istg no one uses this

File: src/lib/qboard.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ import ClipboardHandler from "./clipboard";
 import StyleHandler, { Dash, Fill, Stroke, Style } from "./styles";
 import ActionHandler from "./action";
 import KeyboardHandler, { KeyMap } from "./keyboard";
+import FileHandler from "./files";
```
Comment (cjquines) [inline src/lib/qboard.ts:11]: can we move this before historyhandler? imports are ordered according to how theyre called in constructor xpxpxpxp

**Review by cjquines — CHANGES_REQUESTED** [review]

**Comment by cjquines** [comment]
good idea to factor out file things btw

File: src/lib/files.ts (hunk 2 above)
Comment (pihart) [reply src/lib/files.ts:29]: if you're talking about checking the type, this is how you ensure backwards compatibility when changing your file format.

though i do agree v0 (which doesn't specify qboard-version and is just a pure array) needs to be deprecated

File: src/lib/files.ts [hunk 4]
```diff
@@ -39,9 +39,18 @@ export class JSONReader {
...
-  constructor(private sourceJSON: any[]) {}
+  constructor(pagesJSON: any[]) {
+    this.sourceJSON = {
+      version: 1,
+      pages: pagesJSON,
+    };
+  }
 
   toString = (): string => {
     if (this.stringified) return this.stringified;
```
Comment (pihart) [inline src/lib/files.ts:56]: will never fail in practice but bad

File: src/lib/files.ts (hunk 2 above)
Comment (cjquines) [reply src/lib/files.ts:29]: yeah let's just get rid of v0

**Review by cjquines — CHANGES_REQUESTED** [review]

File: src/lib/files.ts (hunk 2 above)
Comment (pihart) [reply src/lib/files.ts:29]: this breaks any files that were previously exported from your demo. is that fine?

Comment (cjquines) [reply src/lib/files.ts:29]: yes

**Review by cjquines — APPROVED** [review]

**Pull request merged**